### ISSUE TITLE
returning newly-created directories

### ISSUE THREAD

**Issue by zkat** [yoshuawuyts/mkdirp#1]
**Choose one:** 🙋 feature request

tl;dr I want to find out what directories `mkdirp` is actually making, so I can apply things like, say, chmod or chown recursively only to the newly-created stuff.

## Expected Behavior

`Result<Option<Path>>`

## Current Behavior

`Result<()>`

## Possible Solution

This would require reimplementing most of this crate to track newly-created directories, and it would be a breaking change because the function signature would change. It would be a low-impact change, though, because no one's using the wrapped `()` return value for anything (or they shouldn't be?)

## Context

I'm working on https://github.com/zkat/cacache-rs/, which is a port of https://npm.im/cacache, and whenever I insert new data into the cache, I need to have the option to `chownr` any new directories and files, in order to drop permissions when operating under `sudo` (this is a thing that npm does to prevent `~/.npm` from being edited as `root` when you do `sudo npm i`).

**Comment by yoshuawuyts** [comment]
@zkat thanks for the detailed issue, and I'm super in favor of this change! I'm not really using this module much currently; would you be willing to open up a patch to make these changes?

**Comment by zkat** [comment]
I'm more than happy to provide a patch, myself! I'll give it a whirl soon. Thanks!
